merlin-server/sql/merlin/functions/public/restore_from_snapshot.sql

```diff
@@ -0,0 +1,105 @@
+create procedure restore_from_snapshot(_plan_id integer, _snapshot_id integer)
+	language plpgsql as $$
+	declare
+		_snapshot_name text;
+		_plan_name text;
+	begin
+		-- Input Validation
+		select name from plan where id = _plan_id into _plan_name;
+		if _plan_name is null then
+			raise exception 'Cannot Restore: Plan with ID % does not exist.', _plan_id;
+		end if;
+		if not exists(select snapshot_id from plan_snapshot where snapshot_id = _snapshot_id) then
+			raise exception 'Cannot Restore: Snapshot with ID % does not exist.', _snapshot_id;
+		end if;
+		if not exists(select snapshot_id from plan_snapshot where _snapshot_id = snapshot_id and _plan_id = plan_id ) then
+			select snapshot_name from plan_snapshot where snapshot_id = _snapshot_id into _snapshot_name;
+			if _snapshot_name is not null then
+        raise exception 'Cannot Restore: Snapshot ''%'' (ID %) is not a snapshot of Plan ''%'' (ID %)',
+          _snapshot_name, _snapshot_id, _plan_name, _plan_id;
+      else
+				raise exception 'Cannot Restore: Snapshot % is not a snapshot of Plan ''%'' (ID %)',
+          _snapshot_id, _plan_name, _plan_id;
+      end if;
+    end if;
+
+		-- Catch Plan_Locked
+		call plan_locked_exception(_plan_id);
+
+    -- Record the Union of Activities in Plan and Snapshot
+    -- and note which ones have been added since the Snapshot was taken (in_snapshot = false)
+    create temp table diff(
+			activity_id integer,
+			in_snapshot boolean not null
+		);
+		insert into diff(activity_id, in_snapshot)
+		select id as activity_id, true
+		from plan_snapshot_activities where snapshot_id = _snapshot_id;
+
+		insert into diff (activity_id, in_snapshot)
+		select activity_id, false
+		from(
+				select id as activity_id
+				from activity_directive
+				where plan_id = _plan_id
+			except
+				select activity_id
+				from diff) a;
+
+		-- Remove any added activities
+  delete from activity_directive ad
+		using diff d
+		where (ad.id, ad.plan_id) = (d.activity_id, _plan_id)
+			and d.in_snapshot is false;
+
+		-- Upsert the rest
+		insert into activity_directive (
+		      id, plan_id, name, source_scheduling_goal_id, created_at, last_modified_at, last_modified_by,
+		      start_offset, type, arguments, last_modified_arguments_at, metadata,
+		      anchor_id, anchored_to_start)
+		select psa.id, _plan_id, psa.name, psa.source_scheduling_goal_id, psa.created_at, psa.last_modified_at, psa.last_modified_by,
+		       psa.start_offset, psa.type, psa.arguments, psa.last_modified_arguments_at, psa.metadata,
+		       psa.anchor_id, psa.anchored_to_start
+		from plan_snapshot_activities psa
+		where psa.snapshot_id = _snapshot_id
+		on conflict (id, plan_id) do update
+		-- 'last_modified's are skipped during update, as triggers will overwrite them to now()
```

Comment:
This comment is a bit confusing, since we do reference a `last_modified_by` here.

Reply:
Comment has been tweaked